src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,282 @@
+            The data
+        detector_id: str
+            The detector ID, e.g. SPB_IRU_AGIPD1M or SPB_IRU_AGIPD1M
+        groups: int
+            The number of movable groups
+        motors_per_group: int
+            The number of motors per movable group
+        motor_patter: str
+            The pattern to generate motor IDs, expected to have fields:
+            detector_id, q, m
+        data_selector: str
+            The data selector ID
+        """
+        self.detector_id = detector_id
+
+        detector_param = self.DETECTORS.get(detector_id, {})
+        for key in self.ARGS:
+            value = locals().get(key)
+            value = detector_param.get(key, value)
+            setattr(self, key, value)
+
+        args = (self.num_groups, self.num_motors, self.motor_pattern)
+        if any(arg is None for arg in args):
+            raise ValueError(
+                "You need specify all parameters for custom detector")
+
+        self.num_sources = self.num_groups * self.num_motors
+
+        try:
+            self.dc = dc.select(
+                [(source_id, "*") for source_id in self.motor_ids],
+                require_all=True
+            )
+            self.get_key = self._get_key_from_dc
+        except ValueError:
+            if self.data_selector is None:
+                raise ValueError("The motors are not found")
+
+            self.dc = dc.select(self.data_selector)
+            self.get_key = self._get_key_from_data_selector
+
+            # check the keys of data selector
+            data_selector_src = self.dc[self.data_selector]
+            camelcase = [mangle_device_id_camelcase(source_id)
+                         for source_id in self.motor_ids]
+            underscore = [mangle_device_id_underscore(source_id)
+                          for source_id in self.motor_ids]
+
+            if check_data_selector_nodes(data_selector_src, camelcase):
+                self._mangle_id = mangle_device_id_camelcase
+            elif check_data_selector_nodes(data_selector_src, underscore):
+                self._mangle_id = mangle_device_id_underscore
+            else:
+                raise ValueError("The motors are not found")
+
+        self.num_trains = len(self.train_ids)
+
+    def _get_key_from_dc(self, source_id, key):
+        return self.dc[source_id, key]
+
+    def _get_key_from_data_selector(self, source_id, key):
+        data_selector_key = f"{self._mangle_id(source_id)}.{key}"
+        return self.dc[self.data_selector, data_selector_key]
+
+    @property
+    def motor_ids(self):
+        """The list of motor device IDs."""
+        if not hasattr(self, "_motor_ids"):
+            self._motor_ids = [
+                self.motor_pattern.format(
+                    detector_id=self.detector_id,
+                    q=i // self.num_motors + 1,
+                    m=i % self.num_motors + 1,
+                )
+                for i in range(self.num_sources)
+            ]
```

Comment:
Since this is pretty cheap to compute I think it's a bit cleaner to put this in `__init__()`, same for the other places where the `hasattr(...)` pattern is used.

Reply:
This one for sure, thanks. But others methods read the data. This is not necessary cheap. From another side, if we instantiate the component, it implies reading. Should it read in constructor? I think it is not expected in general...

Reply:
Yeah you're right, NVM about the others then.